### PR TITLE
[Tizen] Add a tizen docker image for automatic tests with Qemu

### LINKED ISSUE #22478: [Tizen] Tizen currently does not have automatic functional tests for Matter applications.
### Reproduction steps / Feature

```bash
Tizen does not have automatic functional tests for Matter applications.
```

### Platform

other

### Platform Version(s)

_No response_

### Type

Manually tested with SDK

### (Optional) If manually tested please explain why this is only manually tested

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by DamMicSzm**
#### Problem
Tizen currently does not have automatic functional tests for Matter applications. Resolve #22478

#### Change overview
A docker image was added to prepare a Tizen image for testing in Qemu and automation scripts were created.
For this pull request there will be an additional PR that adds github action to automatic tests

#### Testing
For testing, build a docker image.
```sh
integrations/docker/images/chip-build-tizen-qemu/build.sh 
```
Then build the following applications
```sh
./scripts/build/build_examples.py --target tizen-arm-light-no-ble  --enable-flashbundle build
./scripts/build/build_examples.py --target tizen-arm-chip-tool-no-ble  --enable-flashbundle build
```
Run the docker image with the mounted connectedhomeip project directory and execute the command in project directory.
```sh
./src/test_driver/tizen/run_qemu_test.py --verbose
```

### CODE REVIEW

File: integrations/docker/images/chip-build-tizen-qemu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    && export PKG_CONFIG_PATH="${TIZEN_SDK_SYSROOT}/usr/lib/pkgconfig" \
+    && export PKG_CONFIG_SYSROOT_DIR="${TIZEN_SDK_SYSROOT}" \
+    && ./bootstrap \
+    && ./configure --host=arm-linux-gnueabi --disable-client --disable-manpages --enable-testing --enable-tools \
+    && make lib/bluetooth/bluetooth.h lib/bluetooth/hci.h \
+    && make emulator/btvirt -j$(nproc) \
+    && mkdir temp_btvirt \
+    && install emulator/btvirt temp_btvirt \
+    && guestfish --rw -a $TIZEN_FS_IMAGE_ROOT -m /dev/sda glob copy-in \
+    ./temp_btvirt/* /usr/bin/ \
```
Comment (Arkq) [inline integrations/docker/images/chip-build-tizen-qemu/Dockerfile:126]: Is it required to use `install` command before copying file to FS?

```suggestion
    && guestfish --rw -a $TIZEN_FS_IMAGE_ROOT -m /dev/sda glob copy-in \
    emulator/btvirt /usr/bin/ \
```

File: integrations/docker/images/chip-build-tizen-qemu/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    # Grant access to vHCI for network_fw group
+    && UDEV_BT_VHCI=/usr/lib/udev/rules.d/99-bluetooth-vhci.rules \
+    && echo 'KERNEL=="vhci", MODE="0660", GROUP="network_fw"' > temp.txt \
+    && guestfish --rw -a $TIZEN_FS_IMAGE_ROOT -m /dev/sda upload temp.txt $UDEV_BT_VHCI \
```
Comment (Arkq) [inline integrations/docker/images/chip-build-tizen-qemu/Dockerfile:130]: There is no need to create temp file. Use read from stdin instead (use similar syntax in other placed as well).
```suggestion
    && echo 'KERNEL=="vhci", MODE="0660", GROUP="network_fw"' \
    | guestfish --rw -a $TIZEN_FS_IMAGE_ROOT -m /dev/sda upload - $UDEV_BT_VHCI \
```

File: src/test_driver/tizen/run_qemu_test.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            raise Exception("Can't create ISO file image")
+
+        logging.info("Correct created /opt/app.iso")
+    except:
+        raise
+
+    logging.info("Testing target: %s", target)
+    print("============== TEST BEGIN ================")
+    status = subprocess.Popen('%s/qemu_launcher.sh' % script_path,      text=True,
+                              encoding="850",
```
Comment (Arkq) [inline src/test_driver/tizen/run_qemu_test.py:91]: Why encoding is 850?

File: src/test_driver/tizen/run_qemu_test.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        if verbose:
+            print(line, end='')
+        output += line
+
+    try:
+        if status.returncode != 0:
+            raise Exception("Target %s testing failed with code %d" %
+                            (target, status.returncode))
+
+        for line in output.split('\n'):
```
Comment (Arkq) [inline src/test_driver/tizen/run_qemu_test.py:107]: Maybe this check should be in the `while status.poll() is None` code block?

File: src/test_driver/tizen/run_qemu_test.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            if 'CHIP test' in line and line.find("status: 0") == -1:
```
Comment (Arkq) [inline src/test_driver/tizen/run_qemu_test.py:108]: Use regexp for matching CHIP test status line. Current code might be triggered by something like this:
```py
>>> line = 'Something something CHIP test'
>>> if 'CHIP test' in line and line.find("status: 0") == -1:
...    raise Exception("False positive")

```

**Review by arkq — CHANGES_REQUESTED** [review]

File: src/test_driver/tizen/testing_process.sh [hunk 6]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env bash
```
Comment (andy31415) [inline src/test_driver/tizen/testing_process.sh:1]: can you integrate with build_examples.py ?

if running something like `tizen-tests` it should allow me to run the tests.
Intent is to have a common interface for running things

**Review by andy31415 — CHANGES_REQUESTED** [review]
Please add some documentation on how to use the tizen qemu tests and integrate with build_examples so I can run things like:

```
./scripts/build/build_examples.py --target tizen-qemu-tests build
```

and have tests run.

File: src/test_driver/tizen/run_qemu_test.py (hunk 3 above)
Comment (DamMicSzm) [reply src/test_driver/tizen/run_qemu_test.py:91]: If you don't use this encoding in subprocess. 
```sh
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf1 in position 0: invalid continuation byte.
```
Some line have problem witch decode. When use encoding 850 subprocess and output works corectly.

File: src/test_driver/tizen/testing_process.sh (hunk 6 above)
Comment (DamMicSzm) [reply src/test_driver/tizen/testing_process.sh:1]: Hi @andy31415, thanks for your questions. You're right documentation should be written. I'm have question for you, could you tell me more about why integrations with build_example.py is necessary for you? I'm look on esp32 and there build_example.py build only new target with testing app, not running tests. They have script for running tests something like we have. In our Tizen tests we don't actually have new target testing app like esp32. We only add new docker container with Tizen Qemu image for running examples applications and testing them using "chip-tool" like on physical hardware. For running Qemu and simple tests we wrote python scripts.

Comment (andy31415) [reply src/test_driver/tizen/testing_process.sh:1]: I am generally asking for consistency/convenience. Most developers working on the core SDK are not experts in one particular setup, so reading particular examples for every platform is not convenient. If we have a standard `run esp32-tests` and `run tizen-tests` and such then anyone can just use the vscode docker image and run things. Having custom scripts and platform-specific documentation on the other hand adds friction.

If you just update docker images, I believe this is ok and no integration is needed. If you add a `run_tests.sh` script, this means the tests are scriptable and I am asking for standardization.

Esp32 scripts were created before we tried to standardize compilation instructions, so some things are missing. New things however are encouraged to be uniform from the start.

Comment (Arkq) [reply src/test_driver/tizen/testing_process.sh:1]: > New things however are encouraged to be uniform from the start.

Running tests with the "build_example.py" script is kinda nonintuitive, IMHO (even if there will be a "run" subcommand). Maybe we could add another script for running platform tests? Or maybe already existing `run_test_suite.py` could be adapted to run such tests? Right now Tizen tests are based on a simple script (commissioning + testing light app on/off cluster). In the long run, however, I'd like to run Matter test suite on Tizen (powered by QEMU).

Comment (DamMicSzm) [reply src/test_driver/tizen/testing_process.sh:1]: I understand we need a simpler way to run the tests. I am looking at the existing scripts in the directory ``. / scripts/tests`` and I think the existing scripts can be complicated to integrate with testing using Qemu. 
Do you think it's worth adding a new script to run Qemu-specific tests? Maybe like  ``run_test_qemu.py`` or  ``run_test_qemu.sh`` in directory ``. / scripts/tests``? In this case, any test running in Qemu can be performed using a schema and run from this script. Or maybe you have a better way?

Comment (andy31415) [reply src/test_driver/tizen/testing_process.sh:1]: Resolving: can add this later as things are self-contained inside tizen for now.
I still believe that we should eventually add things into build_examples as an entry point for everything, so we do not need to remember various commands that are platform specific.

**Pull request closed without merging**